src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,378 @@
+ * and never within the actual CPU.
+ *
+ * The I/O for this module is defined in [[AsyncMemIO]].
+ */
+class DualPortedAsyncMemory(size: Int, memfile: String, latency: Int) extends Module {
+  val io = IO(new Bundle {
+    val imem = new AsyncMemIO
+    val dmem = new AsyncMemIO
+  })
+  io := DontCare
+
+  assert(latency > 0) // Check for attempt to make combinational memory
+
+  val memory    = Mem(math.ceil(size.toDouble/4).toInt, UInt(32.W))
+  loadMemoryFromFile(memory, memfile)
+
+  // Instruction port
+
+  val imemPipe = Module(new Pipe(new Request, latency))
+  val imemBusy = Reg(new Bool)
+
+  imemPipe.io := DontCare
+
+  io.imem.request.ready := ~imemBusy
+  when (io.imem.request.ready && io.imem.request.valid
+        && io.imem.request.bits.operation =/= 0.U) {
+    // Put the Request into the instruction pipe and signal that instruction memory is busy
+    val inRequest = io.imem.request.asTypeOf(new Request)
+    imemPipe.io.enq.bits <> inRequest
+    imemPipe.io.enq.valid := true.B
+    imemBusy := true.B
+  }
+ 
+  when (imemPipe.io.deq.valid && imemBusy) {
+    // Dequeue the request out of memory and execute it
+    val outRequest = imemPipe.io.deq.asTypeOf (new Request)
+    when (outRequest.operation === 1.U) { 
+      io.imem.response.valid := true.B
+      io.imem.response.bits  := memory(outRequest.address >> 2)
+    } // Ignore instruction writes as there is no way imem can issue a write access to memory
+    // Signal that instruction memory is idling as we're done
+    imemBusy := false.B
+  }
+
+  // Data port
+
+  val dmemPipe = Module(new Pipe(new Request, latency))
+  val dmemBusy = Reg(new Bool)
+
+  dmemPipe.io := DontCare
+
+  io.dmem.request.ready := ~dmemBusy
+  when (io.dmem.request.ready && io.dmem.request.valid) {
+    // Put the Request into the data pipe and signal that data memory is busy 
+    val inRequest = io.dmem.request.asTypeOf (new Request)
+    when (inRequest.operation =/= 0.U) {
+      dmemPipe.io.enq.valid := true.B
+      dmemPipe.io.enq.bits  := inRequest
+      // Writes do not stall pipeline
+      when (inRequest.operation =/= 1.U) { 
+        dmemBusy := true.B
+      }
+    }
+  }
+ 
+  when (dmemPipe.io.deq.valid) {
+    // Dequeue request and execute
+    val outRequest = dmemPipe.io.deq.asTypeOf (new Request)
+    val address = outRequest.address >> 2
+    // Check that address is pointing to a valid location in memory
+    assert (outRequest.address < size.U)
+    
+    when (outRequest.operation === 1.U) { 
+      io.dmem.response.valid := true.B
+      io.dmem.response.bits  := memory(address)
+    } .elsewhen (outRequest.operation === 2.U) {  
+      memory(address) := outRequest.writedata
+    }
+    // Signal that data memory is idling as we're done
+    when (dmemBusy) {
+      dmemBusy := false.B
+    }
+  }
+}
+
+
+
+
+
+class AsyncMemoryWrapper(size: Int, memFile: String, latency: Int) extends Module {
```

Comment:
I assume this is just for testing? If so, you can put it in the `AsyncMemoryUnitTest.scala` file, IIRC.